File: jvm-tools/interpreters/languages/java-lang-common/src/main/kotlin/interpreters/javalang/common/lexer/nextToken.kt
package interpreters.javalang.common.lexer

import interpreters.javalang.common.tokens.Token
import interpreters.javalang.common.tokens.TokenType
import interpreters.javalang.common.tokens.lookupIdentifier

fun Lexer.nextToken(): Token {
    lateinit var t: Token

    skipWhitespace()

    tokenStartLineNumber = lineNumber()
    tokenStartColumnNumber = columnNumber()

    when (currentCharacter) {
        '=' -> {
            t = if (peekCharacter() == '=') {
                readNextCharacter()
                createNewToken(tokenType = TokenType.DOUBLE_EQUALS, tokenLiteral = TokenType.DOUBLE_EQUALS.value)
            } else {
                createNewToken(tokenType = TokenType.EQUALS, tokenLiteral = TokenType.EQUALS.value)
            }
        }

        ',' -> t = createNewToken(tokenType = TokenType.COMMA, tokenLiteral = TokenType.COMMA.value)
        '+' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '+' -> {
                    readNextCharacter()
                    createNewToken(tokenType = TokenType.INCREMENT, tokenLiteral = TokenType.INCREMENT.value)
                }

                '=' -> {
                    readNextCharacter()
                    createNewToken(tokenType = TokenType.PLUS_EQUALS, tokenLiteral = TokenType.PLUS_EQUALS.value)
                }

                else -> createNewToken(tokenType = TokenType.PLUS, tokenLiteral = TokenType.PLUS.value)
            }
        }

        ';' -> t = createNewToken(tokenType = TokenType.SEMICOLON, tokenLiteral = TokenType.SEMICOLON.value)
        '@' -> t = createNewToken(tokenType = TokenType.AT_SIGN, tokenLiteral = TokenType.AT_SIGN.value)

        '-' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '-' -> {
                    readNextCharacter()
                    createNewToken(tokenType = TokenType.DECREMENT, tokenLiteral = TokenType.DECREMENT.value)
                }

                '=' -> {
                    readNextCharacter()
                    createNewToken(tokenType = TokenType.MINUS_EQUALS, tokenLiteral = TokenType.MINUS_EQUALS.value)
                }

                else -> createNewToken(tokenType = TokenType.MINUS, tokenLiteral = TokenType.MINUS.value)
            }
        }

        '*' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '=' -> {
                    readNextCharacter()
                    createNewToken(
                        tokenType = TokenType.ASTERISK_EQUALS,
                        tokenLiteral = TokenType.ASTERISK_EQUALS.value
                    )
                }

                else -> createNewToken(tokenType = TokenType.ASTERISK, tokenLiteral = TokenType.ASTERISK.value)
            }
        }

        '/' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '=' -> {
                    readNextCharacter()
                    createNewToken(
                        tokenType = TokenType.FORWARD_SLASH_EQUALS,
                        tokenLiteral = TokenType.FORWARD_SLASH_EQUALS.value
                    )
                }

                '/' -> createNewToken(tokenType = TokenType.SINGLE_LINE_COMMENT, tokenLiteral = readSingleLineComment())
                '*' -> createNewToken(tokenType = TokenType.MULTI_LINE_COMMENT, tokenLiteral = readMultilineComment())

                else -> createNewToken(
                    tokenType = TokenType.FORWARD_SLASH,
                    tokenLiteral = TokenType.FORWARD_SLASH.value
                )
            }
        }

        '\\' -> t = createNewToken(tokenType = TokenType.BACK_SLASH, tokenLiteral = TokenType.BACK_SLASH.value)
        '%' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '=' -> {
                    readNextCharacter()
                    createNewToken(tokenType = TokenType.MODULUS_EQUALS, tokenLiteral = TokenType.MODULUS_EQUALS.value)
                }

                else -> createNewToken(tokenType = TokenType.MODULUS, tokenLiteral = TokenType.MODULUS.value)
            }
        }

        '!' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '=' -> {
                    readNextCharacter()
                    createNewToken(tokenType = TokenType.BANG_EQUALS, tokenLiteral = TokenType.BANG_EQUALS.value)
                }

                else -> createNewToken(tokenType = TokenType.BANG, tokenLiteral = TokenType.BANG.value)
            }
        }

        '&' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '&' -> {
                    readNextCharacter()
                    createNewToken(
                        tokenType = TokenType.DOUBLE_AMPERSAND,
                        tokenLiteral = TokenType.DOUBLE_AMPERSAND.value
                    )
                }

                else -> createNewToken(tokenType = TokenType.AMPERSAND, tokenLiteral = TokenType.AMPERSAND.value)
            }
        }

        '|' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '|' -> {
                    readNextCharacter()
                    createNewToken(
                        tokenType = TokenType.DOUBLE_VERTICAL_BAR,
                        tokenLiteral = TokenType.DOUBLE_VERTICAL_BAR.value
                    )
                }

                else -> createNewToken(tokenType = TokenType.VERTICAL_BAR, tokenLiteral = TokenType.VERTICAL_BAR.value)
            }
        }

        '^' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '=' -> {
                    readNextCharacter()
                    createNewToken(tokenType = TokenType.CARET_EQUALS, tokenLiteral = TokenType.CARET_EQUALS.value)
                }

                else -> createNewToken(tokenType = TokenType.CARET, tokenLiteral = TokenType.CARET.value)
            }
        }

        '?' -> t = createNewToken(tokenType = TokenType.QUESTION, tokenLiteral = TokenType.QUESTION.value)
        ':' -> t = createNewToken(tokenType = TokenType.COLON, tokenLiteral = TokenType.COLON.value)
        '.' -> {
            t = if (peekCharacter().isDigit()) {
                readNumberLiteral()
            } else {
                createNewToken(tokenType = TokenType.DOT, tokenLiteral = TokenType.DOT.value)
            }
        }

        '~' -> t = createNewToken(tokenType = TokenType.TILDE, tokenLiteral = TokenType.TILDE.value)
        '\'' -> t = createNewToken(tokenType = TokenType.CHARACTER_LITERAL, tokenLiteral = readCharacterLiteral())
        '"' -> t = createNewToken(tokenType = TokenType.STRING_LITERAL, tokenLiteral = readStringLiteral())
        '(' -> t =
            createNewToken(tokenType = TokenType.LEFT_PARENTHESIS, tokenLiteral = TokenType.LEFT_PARENTHESIS.value)

        ')' -> t =
            createNewToken(tokenType = TokenType.RIGHT_PARENTHESIS, tokenLiteral = TokenType.RIGHT_PARENTHESIS.value)

        '{' -> t =
            createNewToken(tokenType = TokenType.LEFT_CURLY_BRACE, tokenLiteral = TokenType.LEFT_CURLY_BRACE.value)

        '}' -> t =
            createNewToken(tokenType = TokenType.RIGHT_CURLY_BRACE, tokenLiteral = TokenType.RIGHT_CURLY_BRACE.value)

        '[' -> t = createNewToken(
            tokenType = TokenType.LEFT_SQUARE_BRACKET,
            tokenLiteral = TokenType.LEFT_SQUARE_BRACKET.value
        )

        ']' -> t =
            createNewToken(
                tokenType = TokenType.RIGHT_SQUARE_BRACKET,
                tokenLiteral = TokenType.RIGHT_SQUARE_BRACKET.value
            )

        '<' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '<' -> {
                    readNextCharacter()
                    createNewToken(
                        tokenType = TokenType.DOUBLE_LEFT_ANGLE_BRACKET,
                        tokenLiteral = TokenType.DOUBLE_LEFT_ANGLE_BRACKET.value
                    )
                }

                '=' -> {
                    readNextCharacter()
                    createNewToken(
                        tokenType = TokenType.LEFT_ANGLE_BRACKET_EQUALS,
                        tokenLiteral = TokenType.LEFT_ANGLE_BRACKET_EQUALS.value
                    )
                }

                else -> createNewToken(
                    tokenType = TokenType.LEFT_ANGLE_BRACKET,
                    tokenLiteral = TokenType.LEFT_ANGLE_BRACKET.value
                )
            }
        }

        '>' -> {
            val peek = peekCharacter()
            t = when (peek) {
                '>' -> {
                    readNextCharacter()
                    if (peekCharacter() == '>') {
                        readNextCharacter()
                        createNewToken(
                            tokenType = TokenType.TRIPLE_RIGHT_ANGLE_BRACKET,
                            tokenLiteral = TokenType.TRIPLE_RIGHT_ANGLE_BRACKET.value
                        )
                    } else {
                        createNewToken(
                            tokenType = TokenType.DOUBLE_RIGHT_ANGLE_BRACKET,
                            tokenLiteral = TokenType.DOUBLE_RIGHT_ANGLE_BRACKET.value
                        )
                    }
                }

                '=' -> {
                    readNextCharacter()
                    createNewToken(
                        tokenType = TokenType.RIGHT_ANGLE_BRACKET_EQUALS,
                        tokenLiteral = TokenType.RIGHT_ANGLE_BRACKET_EQUALS.value
                    )
                }

                else -> createNewToken(
                    tokenType = TokenType.RIGHT_ANGLE_BRACKET,
                    tokenLiteral = TokenType.RIGHT_ANGLE_BRACKET.value
                )
            }
        }

        EOF_CHARACTER -> t = Token.buildEOF(
            lineNumber = tokenStartLineNumber!!,
            columnNumber = tokenStartColumnNumber!!,
        )

        else -> {
            t = if (currentCharacter.isAcceptableIdentifierStart()) {
                val identifier = readIdentifier()
                // this return is necessary to avoid the unnecessary readNextCharacter
                // call after when block
                return createNewToken(
                    tokenType = lookupIdentifier(identifier),
                    tokenLiteral = identifier,
                )
            } else if (currentCharacter.isDigit()) {
                readNumberLiteral()
            } else {
                createNewToken(tokenType = TokenType.ILLEGAL, tokenLiteral = "$currentCharacter")
            }
        }
    }

    readNextCharacter()

    tokenStartLineNumber = null
    tokenStartColumnNumber = null

    return t
}

File: deno-tools/tools/src/compiler_fe/lexer_xml/next_token.ts
import {
  EOF_CHARACTER,
  type Lexer,
  LexerError,
  type Token,
} from "@/compiler_fe/lexer_core/mod.ts";
import { skipWhitespace } from "./skip_whitespace.ts";
import { readStringLiteral } from "./read_string_literal.ts";
import {
  isAcceptableIdentifierStart,
  readIdentifier,
} from "./read_identifier.ts";
import { readComment } from "./read_comment.ts";
import { readTextNode } from "./read_text_node.ts";
import { XmlTokenType } from "./tokens.ts";

export function nextToken(lexer: Lexer<XmlTokenType>): Token<XmlTokenType> {
  let t: Token<XmlTokenType>;

  skipWhitespace(lexer);

  const currCh = lexer.inputReader.currentCharacter;
  switch (currCh) {
    case ":": {
      t = createToken(lexer, XmlTokenType.Colon);
      break;
    }
    case "=": {
      t = createToken(lexer, XmlTokenType.Equals);
      break;
    }
    case "<": {
      if (lexer.inputReader.peekCharacter() === "!") {
        const { lineNumber, columnNumber } = lexer.inputReader;
        const commentLiteral = readComment(lexer);
        if (commentLiteral !== null) {
          t = createToken(
            lexer,
            XmlTokenType.CommentLiteral,
            commentLiteral,
            lineNumber,
            columnNumber,
          );
        } else {
          t = createToken(
            lexer,
            XmlTokenType.Illegal,
            `${lexer.inputReader.currentCharacter}`,
          );
        }
      } else {
        t = createToken(lexer, XmlTokenType.LeftAngleBracket);
      }
      break;
    }
    case ">": {
      t = createToken(lexer, XmlTokenType.RightAngleBracket);
      break;
    }
    case "/": {
      t = createToken(lexer, XmlTokenType.ForwardSlash);
      break;
    }
    case "?": {
      t = createToken(lexer, XmlTokenType.QuestionMark);
      break;
    }
    case '"': {
      const { lineNumber, columnNumber } = lexer.inputReader;
      const s = readStringLiteral(lexer);
      t = createToken(
        lexer,
        XmlTokenType.StringLiteral,
        s,
        lineNumber,
        columnNumber,
      );
      break;
    }
    case EOF_CHARACTER: {
      t = createToken(lexer, XmlTokenType.Eof);
      break;
    }
    default: {
      if (
        lexer.currentToken == null ||
        lexer.currentToken.tokenType === XmlTokenType.RightAngleBracket
      ) {
        const { lineNumber, columnNumber } = lexer.inputReader;
        const textNode = readTextNode(lexer);
        t = createToken(
          lexer,
          XmlTokenType.TextNode,
          textNode,
          lineNumber,
          columnNumber,
        );
      } else if (isAcceptableIdentifierStart(currCh)) {
        const { lineNumber, columnNumber } = lexer.inputReader;
        const identifier = readIdentifier(lexer);
        t = createToken(
          lexer,
          XmlTokenType.Identifier,
          identifier,
          lineNumber,
          columnNumber,
        );
      } else {
        lexer.addError(
          new LexerError({
            errorCategory: {
              type: "ILLEGAL_CHARACTER",
              ch: currCh,
            },
            lineNumber: lexer.inputReader.lineNumber,
            columnNumber: lexer.inputReader.columnNumber,
          }),
        );
        t = createToken(lexer, XmlTokenType.Illegal, currCh);
      }
      break;
    }
  }

  lexer.inputReader.readNextCharacter();

  lexer.currentToken = t;
  return t;
}

function createToken(
  lexer: Lexer<XmlTokenType>,
  tokenType: XmlTokenType,
  tokenLiteral: string = tokenType.value,
  lineNumber: number = lexer.inputReader.lineNumber,
  columnNumber: number = lexer.inputReader.columnNumber,
): Token<XmlTokenType> {
  return {
    tokenType: tokenType,
    tokenLiteral: tokenLiteral,
    lineNumber,
    columnNumber,
  };
}
